src/Core/OrganizationFeatures/OrganizationUsers/UpdateOrganizationUserCommand.cs

```diff
@@ -0,0 +1,94 @@
+    private readonly ICountNewSmSeatsRequiredQuery _countNewSmSeatsRequiredQuery;
+    private readonly IUpdateSecretsManagerSubscriptionCommand _updateSecretsManagerSubscriptionCommand;
+
+    public UpdateOrganizationUserCommand(
+        IEventService eventService,
+        IOrganizationService organizationService,
+        IOrganizationRepository organizationRepository,
+        IOrganizationUserRepository organizationUserRepository,
+        ICountNewSmSeatsRequiredQuery countNewSmSeatsRequiredQuery,
+        IUpdateSecretsManagerSubscriptionCommand updateSecretsManagerSubscriptionCommand)
+    {
+        _eventService = eventService;
+        _organizationService = organizationService;
+        _organizationRepository = organizationRepository;
+        _organizationUserRepository = organizationUserRepository;
+        _countNewSmSeatsRequiredQuery = countNewSmSeatsRequiredQuery;
+        _updateSecretsManagerSubscriptionCommand = updateSecretsManagerSubscriptionCommand;
+    }
+
+    public async Task UpdateUserAsync(OrganizationUser user, Guid? savingUserId,
+        IEnumerable<CollectionAccessSelection> collections,
+        IEnumerable<Guid> groups)
+    {
+        if (user.Id.Equals(default(Guid)))
+        {
+            throw new BadRequestException("Invite the user first.");
+        }
+
+        var originalUser = await _organizationUserRepository.GetByIdAsync(user.Id);
+        if (user.Equals(originalUser))
+        {
+            throw new BadRequestException("Please make changes before saving.");
+        }
+
+        if (savingUserId.HasValue)
+        {
+            await _organizationService.ValidateOrganizationUserUpdatePermissions(user.OrganizationId, user.Type, originalUser.Type, user.GetPermissions());
+        }
+
+        await _organizationService.ValidateOrganizationCustomPermissionsEnabledAsync(user.OrganizationId, user.Type);
+
+        if (user.Type != OrganizationUserType.Owner &&
+            !await _organizationService.HasConfirmedOwnersExceptAsync(user.OrganizationId, new[] { user.Id }))
+        {
+            throw new BadRequestException("Organization must have at least one confirmed owner.");
+        }
+
+        // Only autoscale (if required) after all validation has passed so that we know it's a valid request before
+        // updating Stripe
+        if (!originalUser.AccessSecretsManager && user.AccessSecretsManager)
+        {
+            var additionalSmSeatsRequired = await _countNewSmSeatsRequiredQuery.CountNewSmSeatsRequiredAsync(user.OrganizationId, 1);
+            if (additionalSmSeatsRequired > 0)
+            {
+                var organization = await _organizationRepository.GetByIdAsync(user.OrganizationId);
+                var update = new SecretsManagerSubscriptionUpdate(organization, true);
+                update.AdjustSeats(additionalSmSeatsRequired);
+                await _updateSecretsManagerSubscriptionCommand.UpdateSubscriptionAsync(update);
+            }
+        }
```

Comment:
This PR mostly looks like a cut & paste, which generally works fine, except here. Commands shouldn't call queries or other commands. I think this means that it needs to be extracted out to the controller level, which calls the commands/queries in order.

This is how the service account code works, which is similar: https://github.com/bitwarden/server/blob/master/src/Api/SecretsManager/Controllers/ServiceAccountsController.cs#L123-L134

The difference here is that both our commands have validation requirements that should be checked first. We don't want to update the orgUser without knowing we can scale the subscription, and vice versa. This is why we have `UpdateSecretsManagerSubscriptionCommand.ValidateAsync` - maybe call that first, then update the orgUser, then call `UpdateSubscriptionAsync`. Or add your own `ValidateAsync` interface to this command if there's a reason to do it the other way around.

Note that the public API doesn't support updating this property, so it doesn't need this code.

Reply:
Excellent point! I forgot that dependencies between commands/queries should not happen.
I've added a few unit tests for `OrganizationUsersController` to be sure the logic was correct.